# Examen A Módulo 1 - Optimización.

**Lea cuidadosamente las siguientes indicaciones antes de comenzar el examen:**
- Para resolver el examen edite este mismo archivo y renómbrelo de la siguiente manera: *Examen1_ApellidoNombre*, donde *ApellidoNombre* corresponde a su apellido paterno con la inicial en mayúscula, seguido de su primer nombre con la inicial en mayúscula **sin tildes**. Por ejemplo, en mi caso el nombre del archivo sería *Examen1_MoralesGabriel*.
- Resuelva los puntos en el espacio provisto. Si requiere agregar más celdas para código o escritura, hágalo (`Esc`,`B`) o (`Esc`,`A`).
- Recuerde que también se está evaluando su capacidad de interpretar los resultados. Escriba sus interpretaciones/conclusiones en celdas utilizando *Markdown*.
- El formato de presentación de todo el examen en general debe ser adecuado. Use tamaños de letra, colores, etiquetas, etcétera.
- No se resuelven dudas de ningún tipo por los profesores de la asignatura. Por favor, absténgase de preguntar.
- Revise bien el tipo de examen. Sólo se le calificará si realiza el tipo de examen asignado.

## 1. Optimización de funciones con `Sympy`

Funciones permitidas: subs, solve, var, expresiones matemáticas


Un cine cobra 86 pesos la entrada estándar y 60 pesos la entrada a estudiantes. Una sala con capacidad para 120 personas se ocupó al 70\%. Los ingresos para esa función fueron de 6,600 MXN. ¿Cuántos estudiantes entraron?

In [1]:
#Ecuaciones
#86x+60y=6600
#x+y=84

import numpy as np
import scipy.optimize as opt
import sympy as sp

sp.var("x")
sp.var("y")

equ1=86*x+60*y-6600
equ2=x+y-84

xequ1=sp.solve(equ1,x)[0]
yequ2=equ2.subs(x,xequ1)
ysol=sp.solve(yequ2,y)[0]
xsol=xequ1.subs(y,ysol)

xsol,ysol

(60, 24)

Entraron 24 estudiantes.

El cine ofrece combos y opciones individuales. El combo 'amigos' tiene un costo de 156 MXN y consiste de 2 refrescos y unas palomitas. Existe también el combo 'nachos' que tiene un costo de 112 MXN y consiste de un refresco y una orden de nachos. Por separado el refresco tiene un costo de 60 MXN, las palomitas 58 MXN y los nachos 72 MXN. Se obtuvo un total de 6450 de ingresos en dulcería:

- 38 combos vendidos
- 70 refrescos vendidos
- 33 órdenes de palomitas grandes vendidas
- 16 órdenes de nachos vendidas

¿Cuántas personas compraron palomitas sueltas?

In [4]:
import numpy as np
import scipy.optimize as opt
import sympy as sp

sp.var("x")#Palomitas
sp.var("y")#refrescos
sp.var("z")#nachos
sp.var("ca")#combo amigos
sp.var("cn")#combo nachos

combos = ca + cn - 38  # Total de combos
eq1 = 2*ca + cn + y - 70  # Total de refrescos
eq2 = ca + x - 33  # Total de palomitas
eq3 = cn + z - 16  # Total de nachos
Total = 156*ca + 112*cn + 60*y + 58*x + 72*z - 6450  # Total de ingresos

resultados=sp.solve((combos,eq1,eq2,eq3,Total),(ca,cn,x,y,z))
xsol=resultados[x]
ysol=resultados[y]
zsol=resultados[z]
casol=resultados[ca]
cnsol=resultados[cn]


xsol,ysol,zsol,casol,cnsol

(5, 4, 6, 28, 10)

Se compraron 5 palomitas sueltas.

## 2. Programación lineal

Una empresa produce costales de 3 tipos de fertilizantes para diferentes aplicaciones utilizando cuantro componentes químicos: azufre, potasio, fósforo y nitrógeno. La información de cada tonelada de fertilizante es:


<table>
  <tr>
      <th>Fertilizante</th>
      <th>Fósforo</th>
      <th>Azufre</th>
      <th>Potasio</th>
      <th>Nitrógeno</th>
      <th>Precio de venta </th>
  </tr>
  <tr>
      <td>A</td>
      <td>3 kg</td>
      <td>5 kg</td>
      <td>4 kg</td>
      <td>2 kg </td>
      <td>\$40000</td>
  </tr>
  <tr>
      <td>B</td>
      <td>4 kg</td>
      <td>3 kg</td>
      <td>2 kg</td>
      <td>1 kg</td>
      <td>\$50000</td>
  </tr>
  <tr>
      <td>C</td>
      <td>5 kg</td>
      <td>2 kg</td>
      <td>3 kg</td>
      <td>4 kg</td>
      <td>\$60000</td>
  </tr>

</table>

La máquina que se usa para mezclar el fertilizante C necesita mantenimiento y por lo tanto sólo puede mezclar 10 toneladas. Además, hay un pedido de 8 toneladas del fertilizante A.

Dado que en almacén se tienen 90 kg de fósforo, 120 kg de nitrógeno, 80 kg de azufre y 95 kg de potasio, ¿cuánto recomiendas que se mezcle de cada fertilizante? (Nota: No se permiten fracciones de toneladas)

$$ \min_{A,B,C} -40000A-50000B-60000C $$
$$ \text {s.t.} $$
$$ 3A+4B+5C≤90 $$
$$ 5A+3B+2C≤80 $$
$$ 4A+2B+3C≤95 $$
$$ 2A+1B+4C≤120 $$
$$ C≤10 $$
$$ A\geq 8 $$

In [19]:
import numpy as np
import scipy.optimize as opt

c=np.array([
    -40000,
    -50000,
    -60000
])

Aub=np.array([
    [3,4,5], #fósforo por cada fertilizante
    [5,3,2],#azufre por cada fertilizante
    [4,2,3],#potasio por cada fertilizante
    [2,1,4],#nitrógeno por cada fertilizante
    [0,0,1], #limitante del fertilizante C
    [-1,0,0] #Pedido fertilizante A 
])

bub=np.array([
    90, #almacén fósforo
    80,#almacén azufre
    95,#almacén potasio
    120,#almacén nitrogéno
    10, #Solo se puede hacer 10 de C
    -8 #Se necesita más de 8 de A
])

sol=opt.linprog(c,Aub,bub, integrality=1)
sol

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -1130000.0
              x: [ 8.000e+00  9.000e+00  6.000e+00]
            nit: -1
          lower:  residual: [ 8.000e+00  9.000e+00  6.000e+00]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  1.000e+00  2.700e+01  7.100e+01
                              4.000e+00  0.000e+00]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
 mip_node_count: 1
 mip_dual_bound: -1130000.0
        mip_gap: 0.0

Se recomienda mezclar 8 del A, 9 del B y 6 del C.

¿Cuál es el ingreso total?

In [24]:
sol.fun

-1130000.0

El ingreso fue de 1130000.

¿Cuánto fósforo quedó en almacén?

In [29]:
Aub@sol.x

array([90., 79., 68., 49.,  6., -8.])

Se usó 90, quedó 0 en el almacén.

## 3. Regresión lineal

Se quiere predecir la resistencia a la compresión de concreto en función de las proporciones de los componentes en la mezcla (cemento, arena, agua, agregado). La constructora que está realizando el análisis tiene datos históricos donde se registraron cantidades de materia y la resistencia medida:

In [35]:
import pandas as pd
data = {
    'Cemento (kg)': [350, 400, 300, 450, 380, 410, 320, 370, 390, 360, 
                     340, 430, 315, 405, 425, 345, 395, 355, 440, 310, 
                     385, 370, 425, 335],
    'Arena (kg)': [600, 550, 700, 500, 580, 540, 650, 590, 570, 610, 
                   630, 520, 660, 545, 530, 620, 560, 605, 515, 670, 
                   575, 590, 535, 625],
    'Agua (kg)': [180, 160, 200, 170, 175, 165, 185, 172, 170, 190, 
                  182, 165, 195, 175, 168, 180, 175, 185, 162, 198, 
                  178, 172, 170, 180],
    'Agregado (kg)': [1000, 1100, 950, 1200, 1050, 1150, 980, 1020, 1080, 970, 
                      990, 1120, 950, 1060, 1140, 990, 1040, 1010, 1160, 940, 
                      1070, 1030, 1110, 1025],
    'Resistencia a la compresión (MPa)': [35, 42, 30, 50, 38, 44, 32, 36, 40, 34, 
                                          33, 48, 31, 43, 47, 37, 41, 35, 49, 29, 
                                          39, 36, 45, 33]
}
data = pd.DataFrame(data)
data

Cemento (kg)  Arena (kg)  Agua (kg)  Agregado (kg)  \
0            350         600        180           1000   
1            400         550        160           1100   
2            300         700        200            950   
3            450         500        170           1200   
4            380         580        175           1050   
5            410         540        165           1150   
6            320         650        185            980   
7            370         590        172           1020   
8            390         570        170           1080   
9            360         610        190            970   
10           340         630        182            990   
11           430         520        165           1120   
12           315         660        195            950   
13           405         545        175           1060   
14           425         530        168           1140   
15           345         620        180            990   
16           395         560        175           1040   
17           355         605        185           1010   
18           440         515        162           1160   
19           310         670        198            940   
20           385         575        178           1070   
21           370         590        172           1030   
22           425         535        170           1110   
23           335         625        180           1025   

    Resistencia a la compresión (MPa)  
0                                  35  
1                                  42  
2                                  30  
3                                  50  
4                                  38  
5                                  44  
6                                  32  
7                                  36  
8                                  40  
9                                  34  
10                                 33  
11                                 48  
12                                 31  
13                                 43  
14                                 47  
15                                 37  
16                                 41  
17                                 35  
18                                 49  
19                                 29  
20                                 39  
21                                 36  
22                                 45  
23                                 33

Ajusta un modelo utilizando mínimos cuadrados para encontrar una función que pueda predecir la resistencia del concreto.

In [38]:
import numpy as np
import scipy.optimize as opt

y=data["Resistencia a la compresión (MPa)"].values
x1=data["Cemento (kg)"].values
x2=data["Arena (kg)"].values
x3=data["Agua (kg)"].values
x4=data["Agregado (kg)"].values
n=len(y)

ones=np.ones([n,1])
x1_factor=np.reshape(x1,[n,1])
x2_factor=np.reshape(x2,[n,1])
x3_factor=np.reshape(x3,[n,1])
x4_factor=np.reshape(x4,[n,1])
X=np.hstack((ones,x1_factor,x2_factor,x3_factor,x4_factor))
y_vector=np.reshape(y,[n,1])

n,p=X.shape
beta=np.random.rand(p)

def funcion(beta,X,y):
  beta=np.reshape(beta,[p,1])
  y_predi=X@beta
  error=y_predi-y
  rss=error.T@error
  return rss

rss=opt.minimize(funcion,beta,args=(X,y_vector))
rss

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 29.84989167470271
        x: [-5.576e+01  1.374e-01  2.436e-02  1.949e-02  2.397e-02]
      nit: 11
      jac: [ 0.000e+00  9.537e-07  1.431e-06  0.000e+00  2.861e-06]
 hess_inv: [[ 4.516e+02 -5.329e-01 ...  1.504e-01 -3.506e-03]
            [-5.329e-01  8.619e-04 ... -4.447e-04 -9.386e-05]
            ...
            [ 1.504e-01 -4.447e-04 ...  1.186e-03  8.110e-05]
            [-3.506e-03 -9.386e-05 ...  8.110e-05  4.315e-05]]
     nfev: 120
     njev: 20

¿Qué efecto tiene cada factor en la resistencia de la mezcla?

In [42]:
rss.x

array([-5.57585005e+01,  1.37376118e-01,  2.43566971e-02,  1.94852003e-02,
        2.39738319e-02])

El cemento, arena, agua y agregado aumentan la resistencia, siendo el cemento el que más tiene efecto.

¿Cuál será la resistencia si mezclo 390 kg de cemento, 570 kg de arena, 170 kg de agua y 1080 kg de agregado?

In [47]:
resis=-5.57856149e+01+390*1.37375856e-01+ 570*2.43564664e-02+ 170*1.94852759e-02+ 1080 *2.39738302e-02
resis

40.878388306999994

Sería de 40.88.

## 4. Regresión logística

Un equipo de ingenieros quiere saber si una muestra de concreto hidráulico fallará o no bajo determinadas condiciones, basándose en experiencias pasadas.

<table>
  <thead>
    <tr>
      <th>Resistencia a la compresión (MPa)</th>
      <th>Edad del concreto (días)</th>
      <th>Relación agua-cemento</th>
      <th>Falla (Sí/No)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>25</td>
      <td>28</td>
      <td>0.45</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>40</td>
      <td>56</td>
      <td>0.35</td>
      <td>No</td>
    </tr>
    <tr>
      <td>30</td>
      <td>28</td>
      <td>0.50</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>50</td>
      <td>90</td>
      <td>0.30</td>
      <td>No</td>
    </tr>
    <tr>
      <td>35</td>
      <td>28</td>
      <td>0.40</td>
      <td>Sí</td>
    </tr>
    <tr>
      <td>55</td>
      <td>120</td>
      <td>0.32</td>
      <td>No</td>
    </tr>
    <tr>
      <td>45</td>
      <td>60</td>
      <td>0.36</td>
      <td>No</td>
    </tr>
    <tr>
      <td>20</td>
      <td>7</td>
      <td>0.55</td>
      <td>Sí</td>
    </tr>
  </tbody>
</table>


1. ¿Qué factor afecta más? (Utiliza un vector de ceros como tu beta inicial)

In [54]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.optimize as opt

Resistencia = [25, 40, 30, 50, 35, 55, 45, 20]
Edad = [28, 56, 28, 90, 28, 120, 60, 7]
AguaCemento = [0.45, 0.35, 0.50, 0.30, 0.40, 0.32, 0.36, 0.55]
Falla = [1, 0, 1, 0, 1, 0, 0, 1]

y = np.array(Falla)
x1 = np.array(Resistencia)
x2 = np.array(Edad)
x3 = np.array(AguaCemento)

n = len(y)
ones = np.ones((n, 1)) 
x1factor = np.reshape(x1, [n, 1])
x2factor = np.reshape(x2, [n, 1])
x3factor = np.reshape(x3, [n, 1])
X = np.hstack((ones, x1factor, x2factor, x3factor))
y_vector = np.reshape(y, [n, 1]) 

def prediccion(beta, X):
    return 1 / (1 + np.exp(-X @ beta))
def J(beta, X, y):
    epsilon = 1e-15 
    n, p = X.shape
    beta = np.reshape(beta, [p, 1])
    y_pred = prediccion(beta, X)
    error = y * np.log(y_pred + epsilon) + (1 - y) * np.log(1 - y_pred + epsilon)
    return -np.sum(error)

beta_inicial = ([0,0,0,0])
res = opt.minimize(J, beta_inicial, args=(X, y_vector))
res

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 3.503392539186779e-07
        x: [ 2.150e+01  1.887e+00 -2.121e+00  1.402e+01]
      nit: 27
      jac: [-7.105e-15  3.576e-06  5.811e-06 -2.980e-08]
 hess_inv: [[ 1.936e+06  1.551e+05 -1.794e+05  1.261e+06]
            [ 1.551e+05  1.242e+04 -1.437e+04  1.010e+05]
            [-1.794e+05 -1.437e+04  1.663e+04 -1.169e+05]
            [ 1.261e+06  1.010e+05 -1.169e+05  8.215e+05]]
     nfev: 150
     njev: 30

In [56]:
optimos=res.x

El factor Beta0 afecta más, despues la relación agua-cemento.

2. Calcula la probabilidad de falla para una muestra con las siguientes características:
- Resistencia a la compresión: 38 MPa
- Edad del concreto: 30 días
- Relación agua-cemento: 0.40

In [61]:
probabilidad = np.array([1, 38, 30, 0.40])

probabilidad_falla = prediccion(optimos, probabilidad)
probabilidad_falla

0.9999999999999996

Al redondear da 1, entonces es sí hay probabilidad de falla.